tests/components/switchbot_cloud/test_config_flow.py

```diff
@@ -0,0 +1,112 @@
+"""Test the SwitchBot via API config flow."""
+from unittest.mock import AsyncMock, patch
+
+import pytest
+
+from homeassistant import config_entries
+from homeassistant.components.switchbot_cloud.config_flow import (
+    CannotConnect,
+    InvalidAuth,
+)
+from homeassistant.components.switchbot_cloud.const import DOMAIN, ENTRY_TITLE
+from homeassistant.const import CONF_API_KEY, CONF_API_TOKEN
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResultType
+
+pytestmark = pytest.mark.usefixtures("mock_setup_entry")
+
+
+async def test_form(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
+    """Test we get the form."""
+    result_init = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result_init["type"] == FlowResultType.FORM
+    assert not result_init["errors"]
+
+    with patch(
+        "homeassistant.components.switchbot_cloud.config_flow.SwitchBotAPI.list_devices",
+        return_value=True,
+    ):
+        result_configure = await hass.config_entries.flow.async_configure(
+            result_init["flow_id"],
+            {
+                CONF_API_TOKEN: "test-token",
+                CONF_API_KEY: "test-secret-key",
+            },
+        )
+        await hass.async_block_till_done()
+
+    assert result_configure["type"] == FlowResultType.CREATE_ENTRY
+    assert result_configure["title"] == ENTRY_TITLE
+    assert result_configure["data"] == {
+        CONF_API_TOKEN: "test-token",
+        CONF_API_KEY: "test-secret-key",
+    }
+    mock_setup_entry.assert_called_once()
+
+
+async def test_form_invalid_auth(hass: HomeAssistant) -> None:
+    """Test we handle invalid auth."""
+    result_init = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.switchbot_cloud.config_flow.SwitchBotAPI.list_devices",
+        side_effect=InvalidAuth,
+    ):
+        result_configure = await hass.config_entries.flow.async_configure(
+            result_init["flow_id"],
+            {
+                CONF_API_TOKEN: "test-token",
+                CONF_API_KEY: "test-secret-key",
+            },
+        )
+
+    assert result_configure["type"] == FlowResultType.FORM
```

Comment:
Please make sure all config flow tests end with either an Abort or a create entry. This way we can test the config flow is able to recover from an error.

To make these tests nicer, checkout `@pytest.mark.parametrize`, for an example check lastFM config flow test `test_flow_fails`. With this you can write one test where you test different exceptions.

Reply:
I used @pytest.mark.parametrize, the code is not shorter.

I am sorry, about
> Please make sure all config flow tests end with either an Abort or a create entry.

As it is a form, I do not believe it will end on an entry creation or an abort but only in redisplaying the form with an error message. I am not sure how to make it end on either abort or create entry. Sending correct data and creating an entry or aborting because of duplicate unique id feels like it belongs to another test.

Reply:
What I mean with this is. Make the test raise an error, returning the form, then fill in the form correct to make sure we test that the config flow is able to create an entry after an error happened